pkg/router/controller/status.go

```diff
@@ -0,0 +1,238 @@
+	plugin     router.Plugin
+	client     client.RoutesNamespacer
+	routerName string
+
+	contentionInterval time.Duration
+	expected           *lru.Cache
+}
+
+// NewStatusAdmitter creates a plugin wrapper that ensures every accepted
+// route has a status field set that matches this router. The admitter manages
+// an LRU of recently seen conflicting updates to handle when two router processes
+// with differing configurations are writing updates at the same time.
+func NewStatusAdmitter(plugin router.Plugin, client client.RoutesNamespacer, name string) *StatusAdmitter {
+	expected, _ := lru.New(1024)
+	return &StatusAdmitter{
+		plugin:     plugin,
+		client:     client,
+		routerName: name,
+
+		contentionInterval: 1 * time.Minute,
+		expected:           expected,
+	}
+}
+
+// nowFn allows the package to be tested
+var nowFn = unversioned.Now
+
+// findOrCreateIngress loops through the router status ingress array looking for an entry
+// that matches name. If there is no entry in the array, it creates one and appends it
+// to the array. If there are multiple entries with that name, the first one is
+// returned and later ones are removed. Changed is returned as true if any part of the
+// array is altered.
+func findOrCreateIngress(route *routeapi.Route, name string) (_ *routeapi.RouteIngress, changed bool) {
+	position := -1
+	updated := make([]routeapi.RouteIngress, 0, len(route.Status.Ingress))
+	for i := range route.Status.Ingress {
+		existing := &route.Status.Ingress[i]
+		if existing.RouterName != name {
+			updated = append(updated, *existing)
+			continue
+		}
+		if position != -1 {
+			changed = true
+			continue
+		}
+		updated = append(updated, *existing)
+		position = i
+	}
+	switch {
+	case position == -1:
+		position = len(route.Status.Ingress)
+		route.Status.Ingress = append(route.Status.Ingress, routeapi.RouteIngress{
+			RouterName: name,
+			Host:       route.Spec.Host,
+		})
+		changed = true
+	case changed:
+		route.Status.Ingress = updated
+	}
+	ingress := &route.Status.Ingress[position]
+	if ingress.Host != route.Spec.Host {
+		ingress.Host = route.Spec.Host
+		changed = true
+	}
+	return ingress, changed
+}
+
+// setIngressCondition records the condition on the ingress, returning true if the ingress was changed and
+// false if no modification was made.
+func setIngressCondition(ingress *routeapi.RouteIngress, condition routeapi.RouteIngressCondition) bool {
+	for _, existing := range ingress.Conditions {
+		//existing.LastTransitionTime = nil
+		if existing == condition {
+			return false
+		}
+	}
+	now := nowFn()
+	condition.LastTransitionTime = &now
+	ingress.Conditions = []routeapi.RouteIngressCondition{condition}
+	return true
+}
+
+func ingressConditionTouched(ingress *routeapi.RouteIngress) *unversioned.Time {
+	var lastTouch *unversioned.Time
+	for _, condition := range ingress.Conditions {
+		if t := condition.LastTransitionTime; t != nil {
+			switch {
+			case lastTouch == nil, t.After(lastTouch.Time):
```

Comment:
This appears to be effectively a logical 'or' between the 2 conditions.  Is it preferable in golang to prefer a switch statement to an 'if' conditional?